Lombiq.Tests.UI/Extensions/VisualVerificationUITestContextExtensions.cs

```diff
@@ -0,0 +1,509 @@
+        // https://github.com/Codeuctivity/ImageSharp.Compare/blob/2.0.46/ImageSharpCompare/ImageSharpCompare.cs#L218.
+        var diff = referenceImage
+            .CompareTo(elementImage);
+
+        context.AppendFailureDump(
+            Path.Combine(
+                configuration.DumpFolderName,
+                $"{configuration.DumpFileNamePrefix ?? string.Empty}{configuration.DiffLogFileName}"),
+            @"
+calculated differences:
+    absoluteError={0},
+    meanError={1},
+    pixelErrorCount={2},
+    pixelErrorPercentage={3}",
+            diff.AbsoluteError,
+            diff.MeanError,
+            diff.PixelErrorCount,
+            diff.PixelErrorPercentage);
+
+        comparator(diff);
+    }
+
+    private static void AssertInternal<TValue>(
+        VisualMatchApprovedContext approvedContext,
+        Func<TValue, TValue, bool> comparer,
+        TValue actual,
+        TValue expected,
+        string propertyName,
+        string condition)
+    {
+        if (comparer(actual, expected))
+        {
+            return;
+        }
+
+        throw new VisualVerificationAssertedException(
+            FormatAssertionMessage(
+                approvedContext,
+                actual,
+                expected,
+                propertyName,
+                condition));
+    }
+
+    private static void AssertInternal<TValue>(
+        Func<TValue, TValue, bool> comparer,
+        TValue actual,
+        TValue expected,
+        string propertyName,
+        string condition)
+    {
+        if (comparer(actual, expected))
+        {
+            return;
+        }
+
+        throw new VisualVerificationAssertedException(
+            FormatAssertionMessage(
+                approvedContext: null,
+                actual,
+                expected,
+                propertyName,
+                condition));
+    }
+
+    private static string FormatAssertionMessage<TValue>(
+        VisualMatchApprovedContext approvedContext,
+        TValue actual,
+        TValue expected,
+        string propertyName,
+        string condition)
+    {
+        var message = new StringBuilder()
+            .AppendLine()
+            .AppendLine(
+                CultureInfo.InvariantCulture,
+                $"{propertyName} should be {condition} {expected} but the calculated value is {actual}");
+
+        if (approvedContext != null)
+        {
+            string loadedFrom = null;
+            if (!string.IsNullOrEmpty(approvedContext.ReferenceImagePath))
+            {
+                loadedFrom = $"file: {approvedContext.ReferenceImagePath}";
+            }
+            else if (!string.IsNullOrEmpty(approvedContext.ReferenceResourceName))
+            {
+                loadedFrom = $"embedded resource: {approvedContext.ReferenceResourceName}";
+            }
+
+            if (!string.IsNullOrEmpty(loadedFrom))
+            {
+                message
+                    .AppendLine()
```

Comment:
```suggestion
            .AppendLine()
            .AppendLine("Visual verification failed since the asserted element looks different from the reference image.")
```